middleware/checkAuth.js

```diff
@@ -0,0 +1,85 @@
+const jwt = require('jsonwebtoken');
+const ERROR = require('../constants').ERROR;
+const User = require('../models/User');
+
+const TOKEN_EXPIRED_ERROR = 'TokenExpiredError';
+
+module.exports = async (req, res, next) => {
+  if (req.method === 'OPTIONS') {
+    return next();
+  }
+
+  let user;
+  let accessToken;
+  let decodedAccessToken;
+  let refreshToken;
+  let decodedRefreshToken;
+
+  try {
+    accessToken = req.headers.authorization.split(' ')[1];
+
+    decodedAccessToken = jwt.verify(accessToken, process.env.JWT_SECRET_KEY);
+
+    const { id, nickname, imageUrl } = decodedAccessToken;
+
+    user = await User.findById(id);
+
+    refreshToken = user.refreshToken;
+
+    try {
+      decodedRefreshToken = jwt.verify(user.refreshToken, process.env.JWT_SECRET_KEY);
+
+    } catch (error) {
+      if (error.name = TOKEN_EXPIRED_ERROR) {
+        const newRefreshToken = jwt.sign({
+          id,
+          nickname,
+          imageUrl,
+        }, process.env.JWT_ACCESS_TOKEN_EXPIRES_IN, {
+          expiresIn: Number(process.env.JWT_REFRESH_TOKEN_EXPIRES_IN)
+        });
+
+        user.refreshToken = newRefreshToken;
+        await user.save();
+        req.user = { id, nickname, imageUrl };
+        next();
+      }
+    }
+
+    req.user = { id, nickname, imageUrl };
+    next();
+  } catch (error) {
+
+    if (error.name = TOKEN_EXPIRED_ERROR) {
+      decoded = jwt.verify(accessToken, process.env.JWT_SECRET_KEY, { ignoreExpiration: true });
+
+      const { id, nickname, imageUrl } = decodedAccessToken;
+
+      try {
+        decodedRefreshToken = jwt.verify(user.refreshToken, process.env.JWT_SECRET_KEY);
```

Comment:
`jwt.verify`가 던지는 에러 때문에 try catch 블럭이 많이 중첩된 것 같은데, `jwt.verify`의 콜백함수를 활용해볼 수 있을 것 같습니다.

```suggestion
        jwt.verify(user.refreshToken, process.env.JWT_SECRET_KEY, (error, decoded) => {
          if (error) {
            // 에러 핸들링
          }
        });
```

Reply:
확실히 callback 함수 활용 부분을 간과한 측면이 있네요 현재 callback을 활용해서 try catch 중첩을 줄이는 방향으로 다시 재작성 중입니다.